### PR TITLE
Move test code into testing module

### PR DESCRIPTION
**Pull request by vitarb**
Moving all test code, including test rules and testservice into a test module.

This change also introduces recently developed internal TestWorkflowRule to the public, which can now be used to write unit tests for workflows more easily. See documentation of the TestWorkflowRule for details.

Users are expected to take a test dependency on our new testing module like:
```
    testImplementation group: 'io.temporal', name: 'temporal-testing', version: 'X.Y.Z'
```

### CODE REVIEW

File: temporal-testing/build.gradle [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+    targetCompatibility = JavaVersion.VERSION_1_8
+    withJavadocJar()
+    withSourcesJar()
+}
+
+dependencies {
+    errorproneJavac('com.google.errorprone:javac:9+181-r4173-1')
+    errorprone('com.google.errorprone:error_prone_core:2.4.0')
+
+    api project(':temporal-sdk')
```
Comment (mfateev) [inline temporal-testing/build.gradle:41]: I believe most of the other dependencies are already part of temporal-sdk. Can we remove them from here?

File: temporal-testing/build.gradle [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+    }
+
+    implementation group: 'ch.qos.logback', name: 'logback-classic', version: '1.2.3'
+    implementation group: 'com.googlecode.junit-toolbox', name: 'junit-toolbox', version: '2.4'
+    implementation group: 'junit', name: 'junit', version: '4.13.1'
+}
+
+configurations.all {
+    // Conflicts with logback
+    exclude group: 'org.slf4j', module: 'slf4j-simple'
```
Comment (mfateev) [inline temporal-testing/build.gradle:60]: I don't think the testing framework needs a concrete slf4j implementation.

File: temporal-testing/build.gradle [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+
+license {
+    header rootProject.file('license-header.txt')
+    exclude '**/*.puml'
+}
+
+task initSubmodules(type: Exec) {
+    description = 'Initializes submodules'
+    commandLine 'git', 'submodule', 'init'
+}
```
Comment (mfateev) [inline temporal-testing/build.gradle:71]: testing doesn't use submodules

File: temporal-testing/build.gradle (hunk 1 above)
Comment (vitarb) [reply temporal-testing/build.gradle:41]: We need errorprone here as we are using it directly in that build file. We can probably configure it in the root project instead, but that would take some time.

File: temporal-testing/build.gradle (hunk 2 above)
Comment (vitarb) [reply temporal-testing/build.gradle:60]: Cleaned up.

File: temporal-testing/build.gradle (hunk 3 above)
Comment (vitarb) [reply temporal-testing/build.gradle:71]: Cleaned up.

File: temporal-sdk/src/main/java/io/temporal/internal/statemachines/ActivityStateMachine.puml [hunk 4]
```diff
@@ -50,5 +50,5 @@ COMPLETED --> [*]
 FAILED --> [*]
 TIMED_OUT --> [*]
 CANCELED --> [*]
-center footer Copyright (C) 2020 Temporal Technologies, Inc. All Rights Reserved.
+center footer Copyright (C) 2021 Temporal Technologies, Inc. All Rights Reserved.
```
Comment (mfateev) [inline temporal-sdk/src/main/java/io/temporal/internal/statemachines/ActivityStateMachine.puml:53]: Keep 2020 year: https://stackoverflow.com/a/2391555/1664318

**Review by mfateev — APPROVED** [review]

**Pull request merged**